lib/Solver/Solver.cpp

```diff
@@ -0,0 +1,363 @@
+    break;
+  case feature::Feature::FeatureKind::FK_UNKNOWN:
+    return NOT_SUPPORTED;
+  }
+
+  Solver->push();
+  return Ok();
+}
+
+Result<SolverErrorCode> Z3Solver::addFeature(const string &FeatureName) {
+  if (OptionToVariableMapping.find(FeatureName) !=
+      OptionToVariableMapping.end()) {
+    return ALREADY_PRESENT;
+  }
+  z3::expr Feature = Context.bool_const(FeatureName.c_str());
+  OptionToVariableMapping.insert(
+      std::make_pair(FeatureName, std::make_unique<z3::expr>(Feature)));
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::addFeature(const string &FeatureName,
+                     const std::vector<int64_t> &Values) {
+  if (OptionToVariableMapping.find(llvm::StringRef(FeatureName)) !=
+      OptionToVariableMapping.end()) {
+    return ALREADY_PRESENT;
+  }
+  z3::expr Feature = Context.int_const(FeatureName.c_str());
+  OptionToVariableMapping.insert(
+      std::make_pair(FeatureName, std::make_unique<z3::expr>(Feature)));
+
+  // Add the numeric values as constraints
+  z3::expr Constraint = Context.bool_val(false);
+  for (int64_t Value : Values) {
+    Constraint = Constraint || (Feature == Context.int_val(Value));
+  }
+  Solver->add(Constraint);
+
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::removeFeature(feature::Feature &FeatureToRemove) {
+  return NOT_SUPPORTED;
+}
+
+Result<SolverErrorCode>
+Z3Solver::addRelationship(const feature::Relationship &R) {
+  const auto *Parent = (const feature::Feature *)R.getParent();
+  auto ParentOption = Context.bool_const(Parent->getName().str().c_str());
+  z3::expr_vector V(Context);
+  for (const auto &Child : R.children()) {
+    const auto *ChildFeature = (const feature::Feature *)Child;
+    V.push_back(Context.bool_const(ChildFeature->getName().str().c_str()));
+  }
+  switch (R.getKind()) {
+  case feature::Relationship::RelationshipKind::RK_ALTERNATIVE:
+    Solver->add(z3::implies(ParentOption, z3::atmost(V, 1) & z3::mk_or(V)));
+    break;
+  case feature::Relationship::RelationshipKind::RK_OR:
+    Solver->add(z3::implies(ParentOption, z3::mk_or(V)));
+    break;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode>
+Z3Solver::addConstraint(feature::Constraint &ConstraintToAdd) {
+  SolverConstraintVisitor SCV(this);
+  bool Succ = SCV.addConstraint(&ConstraintToAdd);
+  if (!Succ) {
+    return SolverErrorCode::NOT_SUPPORTED;
+  }
+  return Ok();
+}
+
+Result<SolverErrorCode, std::unique_ptr<bool>>
+Z3Solver::hasValidConfigurations() {
+  if (Solver->check() == z3::sat) {
+    return Ok(std::make_unique<bool>(true));
+  }
+  return Ok(std::make_unique<bool>(false));
+}
+
+Result<SolverErrorCode, std::unique_ptr<vara::feature::Configuration>>
+Z3Solver::getNextConfiguration() {
+  // Add previous configuration as a constraint
+  excludeCurrentConfiguration();
+
+  // Retrieve the next configuration
+  return getCurrentConfiguration();
+}
+
+Result<SolverErrorCode, std::unique_ptr<uint64_t>>
```

Comment:
same here

Reply:
What would be a good alternative here? I can't use uint64_t since it is in conflict with the error code (which is of the same type).

Reply:
The problem is that the compiler can then no longer distinguish whether I'm returning an error code or the actual number of configurations.